File: sing-box-zenproxy/experimental/clashapi/parser/clash.go
package parser

import (
	"encoding/json"
	"fmt"

	"gopkg.in/yaml.v3"
)

func ParseClash(content string) []ProxyConfig {
	var doc map[string]interface{}
	if err := yaml.Unmarshal([]byte(content), &doc); err != nil {
		return nil
	}

	proxiesRaw, ok := doc["proxies"]
	if !ok {
		return nil
	}
	proxies, ok := proxiesRaw.([]interface{})
	if !ok {
		return nil
	}

	var results []ProxyConfig
	for _, p := range proxies {
		proxy, ok := p.(map[string]interface{})
		if !ok {
			continue
		}
		if pc := parseClashProxy(proxy); pc != nil {
			results = append(results, *pc)
		}
	}
	return results
}

func parseClashProxy(proxy map[string]interface{}) *ProxyConfig {
	proxyType := yamlStr(proxy, "type")
	server := yamlStr(proxy, "server")
	port := yamlPort(proxy, "port")
	name := yamlStr(proxy, "name")
	if name == "" {
		name = fmt.Sprintf("%s:%d", server, port)
	}
	if server == "" || port == 0 {
		return nil
	}

	switch proxyType {
	case "vmess":
		return parseClashVMess(proxy, name, server, port)
	case "vless":
		return parseClashVLess(proxy, name, server, port)
	case "trojan":
		return parseClashTrojan(proxy, name, server, port)
	case "ss":
		return parseClashSS(proxy, name, server, port)
	case "hysteria2", "hy2":
		return parseClashHysteria2(proxy, name, server, port)
	case "socks5":
		return parseClashSocks5(proxy, name, server, port)
	case "http":
		return parseClashHTTP(proxy, name, server, port)
	}
	return nil
}

func parseClashVMess(proxy map[string]interface{}, name, server string, port uint16) *ProxyConfig {
	uuid := yamlStr(proxy, "uuid")
	if uuid == "" {
		return nil
	}
	alterID := yamlInt(proxy, "alterId")
	cipher := yamlStrDefault(proxy, "cipher", "auto")

	outbound := map[string]interface{}{
		"type":        "vmess",
		"server":      server,
		"server_port": port,
		"uuid":        uuid,
		"alter_id":    alterID,
		"security":    cipher,
	}

	applyClashTransport(proxy, outbound)
	applyClashTLS(proxy, outbound, server)

	raw, _ := json.Marshal(outbound)
	return &ProxyConfig{Name: name, Type: "vmess", Server: server, Port: port, Outbound: raw}
}

func parseClashVLess(proxy map[string]interface{}, name, server string, port uint16) *ProxyConfig {
	uuid := yamlStr(proxy, "uuid")
	if uuid == "" {
		return nil
	}

	outbound := map[string]interface{}{
		"type":        "vless",
		"server":      server,
		"server_port": port,
		"uuid":        uuid,
	}

	if flow := yamlStr(proxy, "flow"); flow != "" {
		outbound["flow"] = flow
	}

	applyClashTransport(proxy, outbound)
	applyClashTLS(proxy, outbound, server)

	// Reality
	if realityOpts, ok := proxy["reality-opts"].(map[string]interface{}); ok {
		if tlsMap, ok := outbound["tls"].(map[string]interface{}); ok {
			pbk := yamlStr(realityOpts, "public-key")
			sid := yamlStr(realityOpts, "short-id")
			tlsMap["reality"] = map[string]interface{}{
				"enabled":    true,
				"public_key": pbk,
				"short_id":   sid,
			}
		}
	}

	raw, _ := json.Marshal(outbound)
	return &ProxyConfig{Name: name, Type: "vless", Server: server, Port: port, Outbound: raw}
}

func parseClashTrojan(proxy map[string]interface{}, name, server string, port uint16) *ProxyConfig {
	password := yamlStr(proxy, "password")
	if password == "" {
		return nil
	}
	sni := yamlStrDefault(proxy, "sni", server)

	outbound := map[string]interface{}{
		"type":        "trojan",
		"server":      server,
		"server_port": port,
		"password":    password,
		"tls": map[string]interface{}{
			"enabled":     true,
			"server_name": sni,
			"insecure":    true,
		},
	}

	applyClashTransport(proxy, outbound)

	raw, _ := json.Marshal(outbound)
	return &ProxyConfig{Name: name, Type: "trojan", Server: server, Port: port, Outbound: raw}
}

func parseClashSS(proxy map[string]interface{}, name, server string, port uint16) *ProxyConfig {
	cipher := yamlStr(proxy, "cipher")
	password := yamlStr(proxy, "password")
	if cipher == "" || password == "" {
		return nil
	}

	outbound := map[string]interface{}{
		"type":        "shadowsocks",
		"server":      server,
		"server_port": port,
		"method":      cipher,
		"password":    password,
	}

	raw, _ := json.Marshal(outbound)
	return &ProxyConfig{Name: name, Type: "shadowsocks", Server: server, Port: port, Outbound: raw}
}

func parseClashHysteria2(proxy map[string]interface{}, name, server string, port uint16) *ProxyConfig {
	password := yamlStr(proxy, "password")
	if password == "" {
		return nil
	}
	sni := yamlStrDefault(proxy, "sni", server)

	outbound := map[string]interface{}{
		"type":        "hysteria2",
		"server":      server,
		"server_port": port,
		"password":    password,
		"tls": map[string]interface{}{
			"enabled":     true,
			"server_name": sni,
			"insecure":    true,
		},
	}

	if obfs := yamlStr(proxy, "obfs"); obfs == "salamander" {
		obfsPassword := yamlStr(proxy, "obfs-password")
		outbound["obfs"] = map[string]interface{}{
			"type":     "salamander",
			"password": obfsPassword,
		}
	}

	raw, _ := json.Marshal(outbound)
	return &ProxyConfig{Name: name, Type: "hysteria2", Server: server, Port: port, Outbound: raw}
}

func parseClashSocks5(proxy map[string]interface{}, name, server string, port uint16) *ProxyConfig {
	outbound := map[string]interface{}{
		"type":        "socks",
		"server":      server,
		"server_port": port,
		"version":     "5",
	}

	if username := yamlStr(proxy, "username"); username != "" {
		outbound["username"] = username
		outbound["password"] = yamlStr(proxy, "password")
	}

	if yamlBool(proxy, "tls") {
		applyClashTLS(proxy, outbound, server)
	}

	raw, _ := json.Marshal(outbound)
	return &ProxyConfig{Name: name, Type: "socks", Server: server, Port: port, Outbound: raw}
}

func parseClashHTTP(proxy map[string]interface{}, name, server string, port uint16) *ProxyConfig {
	outbound := map[string]interface{}{
		"type":        "http",
		"server":      server,
		"server_port": port,
	}

	if username := yamlStr(proxy, "username"); username != "" {
		outbound["username"] = username
		outbound["password"] = yamlStr(proxy, "password")
	}

	if yamlBool(proxy, "tls") {
		applyClashTLS(proxy, outbound, server)
	}

	raw, _ := json.Marshal(outbound)
	return &ProxyConfig{Name: name, Type: "http", Server: server, Port: port, Outbound: raw}
}

func applyClashTransport(proxy map[string]interface{}, outbound map[string]interface{}) {
	network := yamlStrDefault(proxy, "network", "tcp")
	switch network {
	case "ws":
		wsOpts, _ := proxy["ws-opts"].(map[string]interface{})
		path := "/"
		host := ""
		if wsOpts != nil {
			if p := yamlStr(wsOpts, "path"); p != "" {
				path = p
			}
			if headers, ok := wsOpts["headers"].(map[string]interface{}); ok {
				host = yamlStr(headers, "Host")
			}
		}
		outbound["transport"] = map[string]interface{}{
			"type":    "ws",
			"path":    path,
			"headers": map[string]interface{}{"Host": host},
		}
	case "grpc":
		grpcOpts, _ := proxy["grpc-opts"].(map[string]interface{})
		serviceName := ""
		if grpcOpts != nil {
			serviceName = yamlStr(grpcOpts, "grpc-service-name")
		}
		outbound["transport"] = map[string]interface{}{
			"type":         "grpc",
			"service_name": serviceName,
		}
	case "h2":
		h2Opts, _ := proxy["h2-opts"].(map[string]interface{})
		path := "/"
		host := ""
		if h2Opts != nil {
			if p := yamlStr(h2Opts, "path"); p != "" {
				path = p
			}
			if hostArr, ok := h2Opts["host"].([]interface{}); ok && len(hostArr) > 0 {
				if h, ok := hostArr[0].(string); ok {
					host = h
				}
			}
		}
		outbound["transport"] = map[string]interface{}{
			"type": "http",
			"host": []string{host},
			"path": path,
		}
	}
}

func applyClashTLS(proxy map[string]interface{}, outbound map[string]interface{}, server string) {
	tlsEnabled := yamlBool(proxy, "tls")
	if !tlsEnabled {
		return
	}

	sni := yamlStr(proxy, "servername")
	if sni == "" {
		sni = yamlStr(proxy, "sni")
	}
	if sni == "" {
		sni = server
	}
	skipVerify := true
	if v, ok := proxy["skip-cert-verify"]; ok {
		if b, ok := v.(bool); ok {
			skipVerify = b
		}
	}

	tlsConfig := map[string]interface{}{
		"enabled":     true,
		"server_name": sni,
		"insecure":    skipVerify,
	}

	if fp := yamlStr(proxy, "client-fingerprint"); fp != "" {
		tlsConfig["utls"] = map[string]interface{}{
			"enabled":     true,
			"fingerprint": fp,
		}
	}

	outbound["tls"] = tlsConfig
}

// --- YAML helpers ---

func yamlStr(m map[string]interface{}, key string) string {
	v, ok := m[key]
	if !ok || v == nil {
		return ""
	}
	if s, ok := v.(string); ok {
		return s
	}
	return fmt.Sprintf("%v", v)
}

func yamlStrDefault(m map[string]interface{}, key, def string) string {
	s := yamlStr(m, key)
	if s == "" {
		return def
	}
	return s
}

func yamlPort(m map[string]interface{}, key string) uint16 {
	v, ok := m[key]
	if !ok || v == nil {
		return 0
	}
	switch n := v.(type) {
	case int:
		return uint16(n)
	case float64:
		return uint16(n)
	case string:
		p, _ := fmt.Sscanf(n, "%d")
		return uint16(p)
	}
	return 0
}

func yamlInt(m map[string]interface{}, key string) int {
	v, ok := m[key]
	if !ok || v == nil {
		return 0
	}
	switch n := v.(type) {
	case int:
		return n
	case float64:
		return int(n)
	}
	return 0
}

func yamlBool(m map[string]interface{}, key string) bool {
	v, ok := m[key]
	if !ok || v == nil {
		return false
	}
	if b, ok := v.(bool); ok {
		return b
	}
	return false
}

File: src/parser/clash.rs
use super::{ProxyConfig, ProxyType};
use serde_json::json;

pub fn parse(content: &str) -> Vec<ProxyConfig> {
    let yaml: serde_yaml::Value = match serde_yaml::from_str(content) {
        Ok(v) => v,
        Err(e) => {
            tracing::warn!("Failed to parse Clash YAML: {e}");
            return vec![];
        }
    };

    let proxies = yaml.get("proxies").and_then(|p| p.as_sequence());
    let proxies = match proxies {
        Some(p) => p,
        None => return vec![],
    };

    proxies
        .iter()
        .filter_map(|proxy| parse_clash_proxy(proxy))
        .collect()
}

fn parse_clash_proxy(proxy: &serde_yaml::Value) -> Option<ProxyConfig> {
    let proxy_type = proxy.get("type")?.as_str()?;
    let server = proxy.get("server")?.as_str()?.to_string();
    let port = proxy.get("port")?.as_u64()? as u16;
    let name = proxy
        .get("name")
        .and_then(|n| n.as_str())
        .unwrap_or(&format!("{server}:{port}"))
        .to_string();

    match proxy_type {
        "vmess" => parse_clash_vmess(proxy, &name, &server, port),
        "vless" => parse_clash_vless(proxy, &name, &server, port),
        "trojan" => parse_clash_trojan(proxy, &name, &server, port),
        "ss" => parse_clash_ss(proxy, &name, &server, port),
        "hysteria2" | "hy2" => parse_clash_hysteria2(proxy, &name, &server, port),
        "socks5" => parse_clash_socks(proxy, &name, &server, port),
        "http" => parse_clash_http(proxy, &name, &server, port),
        _ => None,
    }
}

fn parse_clash_vmess(
    proxy: &serde_yaml::Value,
    name: &str,
    server: &str,
    port: u16,
) -> Option<ProxyConfig> {
    let uuid = proxy.get("uuid")?.as_str()?;
    let alter_id = proxy
        .get("alterId")
        .and_then(|a| a.as_u64())
        .unwrap_or(0);
    let cipher = proxy
        .get("cipher")
        .and_then(|c| c.as_str())
        .unwrap_or("auto");

    let mut outbound = json!({
        "type": "vmess",
        "server": server,
        "server_port": port,
        "uuid": uuid,
        "alter_id": alter_id,
        "security": cipher,
    });

    apply_clash_transport(proxy, &mut outbound, server);
    apply_clash_tls(proxy, &mut outbound, server);

    Some(ProxyConfig {
        name: name.to_string(),
        proxy_type: ProxyType::VMess,
        server: server.to_string(),
        port,
        singbox_outbound: outbound,
    })
}

fn parse_clash_vless(
    proxy: &serde_yaml::Value,
    name: &str,
    server: &str,
    port: u16,
) -> Option<ProxyConfig> {
    let uuid = proxy.get("uuid")?.as_str()?;

    let mut outbound = json!({
        "type": "vless",
        "server": server,
        "server_port": port,
        "uuid": uuid,
    });

    if let Some(flow) = proxy.get("flow").and_then(|f| f.as_str()) {
        if !flow.is_empty() {
            outbound["flow"] = json!(flow);
        }
    }

    apply_clash_transport(proxy, &mut outbound, server);
    apply_clash_tls(proxy, &mut outbound, server);

    // Reality
    if let Some(reality_opts) = proxy.get("reality-opts") {
        if let Some(tls) = outbound.get_mut("tls") {
            let pbk = reality_opts
                .get("public-key")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            let sid = reality_opts
                .get("short-id")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            tls["reality"] = json!({
                "enabled": true,
                "public_key": pbk,
                "short_id": sid,
            });
        }
    }

    Some(ProxyConfig {
        name: name.to_string(),
        proxy_type: ProxyType::VLESS,
        server: server.to_string(),
        port,
        singbox_outbound: outbound,
    })
}

fn parse_clash_trojan(
    proxy: &serde_yaml::Value,
    name: &str,
    server: &str,
    port: u16,
) -> Option<ProxyConfig> {
    let password = proxy.get("password")?.as_str()?;
    let sni = proxy
        .get("sni")
        .and_then(|s| s.as_str())
        .unwrap_or(server);

    let mut outbound = json!({
        "type": "trojan",
        "server": server,
        "server_port": port,
        "password": password,
        "tls": {
            "enabled": true,
            "server_name": sni,
            "insecure": true,
        }
    });

    apply_clash_transport(proxy, &mut outbound, server);

    Some(ProxyConfig {
        name: name.to_string(),
        proxy_type: ProxyType::Trojan,
        server: server.to_string(),
        port,
        singbox_outbound: outbound,
    })
}

fn parse_clash_ss(
    proxy: &serde_yaml::Value,
    name: &str,
    server: &str,
    port: u16,
) -> Option<ProxyConfig> {
    let cipher = proxy.get("cipher")?.as_str()?;
    let password = proxy.get("password")?.as_str()?;

    let outbound = json!({
        "type": "shadowsocks",
        "server": server,
        "server_port": port,
        "method": cipher,
        "password": password,
    });

    Some(ProxyConfig {
        name: name.to_string(),
        proxy_type: ProxyType::Shadowsocks,
        server: server.to_string(),
        port,
        singbox_outbound: outbound,
    })
}

fn parse_clash_hysteria2(
    proxy: &serde_yaml::Value,
    name: &str,
    server: &str,
    port: u16,
) -> Option<ProxyConfig> {
    let password = proxy.get("password")?.as_str()?;
    let sni = proxy
        .get("sni")
        .and_then(|s| s.as_str())
        .unwrap_or(server);

    let mut outbound = json!({
        "type": "hysteria2",
        "server": server,
        "server_port": port,
        "password": password,
        "tls": {
            "enabled": true,
            "server_name": sni,
            "insecure": true,
        }
    });

    if let Some(obfs) = proxy.get("obfs").and_then(|o| o.as_str()) {
        let obfs_password = proxy
            .get("obfs-password")
            .and_then(|o| o.as_str())
            .unwrap_or("");
        if obfs == "salamander" {
            outbound["obfs"] = json!({
                "type": "salamander",
                "password": obfs_password,
            });
        }
    }

    Some(ProxyConfig {
        name: name.to_string(),
        proxy_type: ProxyType::Hysteria2,
        server: server.to_string(),
        port,
        singbox_outbound: outbound,
    })
}

fn parse_clash_socks(
    proxy: &serde_yaml::Value,
    name: &str,
    server: &str,
    port: u16,
) -> Option<ProxyConfig> {
    let mut outbound = json!({
        "type": "socks",
        "server": server,
        "server_port": port,
        "version": "5",
    });

    if let Some(username) = proxy.get("username").and_then(|u| u.as_str()) {
        if !username.is_empty() {
            outbound["username"] = json!(username);
            let password = proxy
                .get("password")
                .and_then(|p| p.as_str())
                .unwrap_or("");
            outbound["password"] = json!(password);
        }
    }

    // Some Clash configs use "tls" field for socks5
    let tls_enabled = proxy
        .get("tls")
        .and_then(|t| t.as_bool())
        .unwrap_or(false);
    if tls_enabled {
        apply_clash_tls(proxy, &mut outbound, server);
    }

    Some(ProxyConfig {
        name: name.to_string(),
        proxy_type: ProxyType::Socks,
        server: server.to_string(),
        port,
        singbox_outbound: outbound,
    })
}

fn parse_clash_http(
    proxy: &serde_yaml::Value,
    name: &str,
    server: &str,
    port: u16,
) -> Option<ProxyConfig> {
    let mut outbound = json!({
        "type": "http",
        "server": server,
        "server_port": port,
    });

    if let Some(username) = proxy.get("username").and_then(|u| u.as_str()) {
        if !username.is_empty() {
            outbound["username"] = json!(username);
            let password = proxy
                .get("password")
                .and_then(|p| p.as_str())
                .unwrap_or("");
            outbound["password"] = json!(password);
        }
    }

    let tls_enabled = proxy
        .get("tls")
        .and_then(|t| t.as_bool())
        .unwrap_or(false);
    if tls_enabled {
        apply_clash_tls(proxy, &mut outbound, server);
    }

    Some(ProxyConfig {
        name: name.to_string(),
        proxy_type: ProxyType::Http,
        server: server.to_string(),
        port,
        singbox_outbound: outbound,
    })
}

fn apply_clash_transport(
    proxy: &serde_yaml::Value,
    outbound: &mut serde_json::Value,
    _server: &str,
) {
    let network = proxy
        .get("network")
        .and_then(|n| n.as_str())
        .unwrap_or("tcp");
    match network {
        "ws" => {
            let path = proxy
                .get("ws-opts")
                .and_then(|o| o.get("path"))
                .and_then(|p| p.as_str())
                .unwrap_or("/");
            let host = proxy
                .get("ws-opts")
                .and_then(|o| o.get("headers"))
                .and_then(|h| h.get("Host"))
                .and_then(|h| h.as_str())
                .unwrap_or("");
            outbound["transport"] = json!({
                "type": "ws",
                "path": path,
                "headers": { "Host": host }
            });
        }
        "grpc" => {
            let service_name = proxy
                .get("grpc-opts")
                .and_then(|o| o.get("grpc-service-name"))
                .and_then(|s| s.as_str())
                .unwrap_or("");
            outbound["transport"] = json!({
                "type": "grpc",
                "service_name": service_name
            });
        }
        "h2" => {
            let path = proxy
                .get("h2-opts")
                .and_then(|o| o.get("path"))
                .and_then(|p| p.as_str())
                .unwrap_or("/");
            let host = proxy
                .get("h2-opts")
                .and_then(|o| o.get("host"))
                .and_then(|h| h.as_sequence())
                .and_then(|arr| arr.first())
                .and_then(|h| h.as_str())
                .unwrap_or("");
            outbound["transport"] = json!({
                "type": "http",
                "host": [host],
                "path": path
            });
        }
        _ => {}
    }
}

fn apply_clash_tls(
    proxy: &serde_yaml::Value,
    outbound: &mut serde_json::Value,
    server: &str,
) {
    let tls_enabled = proxy
        .get("tls")
        .and_then(|t| t.as_bool())
        .unwrap_or(false);
    if tls_enabled {
        let sni = proxy
            .get("servername")
            .or_else(|| proxy.get("sni"))
            .and_then(|s| s.as_str())
            .unwrap_or(server);
        let skip_verify = proxy
            .get("skip-cert-verify")
            .and_then(|s| s.as_bool())
            .unwrap_or(true);
        outbound["tls"] = json!({
            "enabled": true,
            "server_name": sni,
            "insecure": skip_verify,
        });

        // UTLS fingerprint
        if let Some(fp) = proxy.get("client-fingerprint").and_then(|f| f.as_str()) {
            outbound["tls"]["utls"] = json!({
                "enabled": true,
                "fingerprint": fp
            });
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_sc0() {
        let content = std::fs::read_to_string("C:/tmp/sc0.yaml").unwrap();
        let result = parse(&content);
        println!("Parsed {} proxies", result.len());
        for p in &result {
            println!("  {} | {} | {}:{}", p.name, p.proxy_type, p.server, p.port);
        }
        assert!(result.len() > 0, "Should parse some proxies");
    }
}
